### DISCUSSION TITLE
What needs to be done to make rune generate C code?

### DISCUSSION THREAD

**Discussion by 0dminnimda** [google/rune#43] [General]
What are the approximate steps that needs to be made to enable rune to generate C code, so C compilers optimisations can be harvested?

**Comment by waywardgeek** [comment]
There are 2 approaches, and whoever builds it will probably decide which path to take.

1) Upgrade the existing C compiler to make writing a C/C++ backend simpler, and then write a new code generator in C, linked into the existing compiler.
2) Wait for the bootstrap compiler to be functional and write the new code generator in Rune.

The current bottleneck is that I'm writing most of the C code myself, since I am the only one with extensive experience writing C code backed by the DataDraw code generator.  Once the new binder is complete (maybe 4 weeks from now?) we should be able to make good progress rewriting Rune in Rune.  Also, other folks at Google and in the community might be in a better position to contribute code in Rune.

The current code generator is a 4,000 line mess that needs rewriting in any case.  I read a tutorial on using LLVM to write a new compiler for a new language, and the tutorial said that after type resolution that writing the LLVM IR was easy, and more or less left it as an exercise for the reader.  The tutorial was wrong!  The right approach is to do all the mods to the AST (or DataDraw classes in our case), such as adding calls to ref/unref for ref counted objects, allocation/freeing of objects in constructors/destructors, allocation/freeing of temporary arrays, etc.  Most of that 4,000 line code generator is devoted to writing LLVM code that works as if those changes had already been applied to the AST.

In short, whoever wants to tackle this first probably gets to decide the approach, and they'll get to decide whether to work in DataDraw style C code vs Rune.

**Reply by 0dminnimda** [reply]
> the tutorial said that after type resolution that writing the LLVM IR was easy

From what I understand working with IR is much more troublesome. It does seems like a bad recommendation

> The right approach is to do all the mods to the AST (or DataDraw classes in our case)

One of the things I was thinking is that it's possible to just generate c code (some optimisations requere platform specific compilers which can go further than clang), although it's not much harder than generateing ast and allows above mentioned modifications and it's pretty easy to dump the ast into c code afterwards

> 2\. Wait for the bootstrap compiler to be functional and write the new code generator in Rune.

What features that are required for making a compiler are not still in rune?

Also I wonder if it's possible to start with partial replacement of compiler c code that can already be replaced by rune
